### ISSUE TITLE
It works with automatic size but crashed with custom size

### ISSUE THREAD

**Issue by BhavinBhadani** [sgr-ksmt/WaterfallLayout#3]
It works with automatic size

`return WaterfallLayout.automaticSize`

But if we give some custom size it crashed on prepare layout

**Comment by koust** [comment]
I get the same error, did you solve the problem?

**Comment by BhavinBhadani** [comment]
@koust nope :(

**Comment by pavankataria** [comment]
bump.

**Comment by macvinod** [comment]
same error getting

**Comment by khawars** [comment]
Facing same issue here.

**Comment by Fuxingloh** [comment]
#10 Fixes it but diable auto sizing for waterfall mode.

**Comment by n1schal** [comment]
I copied the updated code but it's still crashing for me.

let layout = WaterfallLayout()
        layout.delegate = self
        layout.sectionInset = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)
        layout.minimumLineSpacing = 8.0
        layout.minimumInteritemSpacing = 8.0
        layout.headerHeight = AppDimensions.HeaderHeight
        collectionView?.collectionViewLayout = layout`

extension HomeCollectionViewController : WaterfallLayoutDelegate{
    func collectionViewLayout(for section: Int) -> WaterfallLayout.Layout {
        switch collectionView?.numberOfItems(inSection: section) {
        case 1:
            return .flow(column: 1)
        case 2:
            return .flow(column: 2)
        case 3:
           return .waterfall(column: 2, distributionMethod: .balanced)
        default:
            return .flow(column: 2)
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, layout: WaterfallLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        var width: CGFloat!
        var height: CGFloat!
        let picCount = collectionView.numberOfItems(inSection: indexPath.section)
        switch picCount {
        case 1:
            width = scW
            height = scW
        case 2:
            width = (scW-8)/2
            height = scW
        case 3:
            height = scW/2
            if(indexPath.row == 0){
                width = scW
            }else{
                width = (scW-8)/2
            }
        case 4:
            Log.add(info: "Width: \(scW)")
            width = (scW-8)/2
            height = (scW-8)/2
        default:
            width = (scW-8)/2
            height = (scW-8)/2
        }
        return CGSize.init(width: width, height: height)
    }
}
`

scWidth is ScreenWidth from UIScreen.main.bounds.width

**Comment by n1schal** [comment]
The only constraints I have on the imageView inside the cell is 0 on all side i.e. leading, trailing, top, bottom

**Comment by Fuxingloh** [comment]
Try changing their priority to high (750).

**Comment by gabzsa** [comment]
Yeah it has nothing to do with constraints (at least in my case). I have cells that have no constraints at all, and it's crashing in prepareLayout in waterfall mode (EXC_BAD_ACCESS) on line 261

**Comment by Fuxingloh** [comment]
That sucks, but EXC_BAD_ACCESS can be thrown due to constraints issues (unable to satisfy constraints simultaneously). I had the same error before I changed the constraints priority.

If you insist on using this library, I recommend you remove all the constraints for every single cell and slowly introduce them back until you fixed all the cells constraints. Including non waterfall mode cells.

**Comment by gabzsa** [comment]
Like I wrote, no constraints. Also constraint errors cannot cause bad access, ios breaks then if unsolvable.

**Comment by Fuxingloh** [comment]
Oh, what I meant is that because of how this lib implement shouldInvalidateLayout(forPreferredLayoutAttributes... which is false in the default UIKit implementation. It causes a stack overflow event because, both the cells and the collection view is suggesting a different set of constraints via a repeated loop. 

The reasons why iOS doesn't attempt to break if unsolvable is because it's not caused via the AutoLayout Engine but rather this delegate. And an infinite recursion causes EXC_BAD_ACCESS in iOS. 

I hope this helps.

**Comment by yokoboko** [comment]
How can we fix that? If I use autoresize mask it works but with constraint layout it does not. I have simple cell with UIImageView constrained to top, bottom, left & right?

**Comment by iamcam** [comment]
I made some progress on this issue so far. I can reproduce the crash by adding a xib for ColorView and a subView (eg, UIImageView), and setting auto layout constraints. Scrolling the content will soon lead to a BAD ACCESS crash. I started reading through [this SO answer](https://stackoverflow.com/a/25896386/694080). He mentioned adding this code to your collection cell,

```
//forces the system to do one layout pass
var isHeightCalculated: Bool = false

override func preferredLayoutAttributesFittingAttributes(_ layoutAttributes: UICollectionViewLayoutAttributes) -> UICollectionViewLayoutAttributes {
    //Exhibit A - We need to cache our calculation to prevent a crash.
    if !isHeightCalculated {
        setNeedsLayout()
        layoutIfNeeded()
        let size = contentView.systemLayoutSizeFitting(layoutAttributes.size)
        var newFrame = layoutAttributes.frame
        newFrame.size.width = CGFloat(ceilf(Float(size.width)))
        layoutAttributes.frame = newFrame
        isHeightCalculated = true
    }
    return layoutAttributes
}
```

... and so I tried it. The crash stops and the cells appear to be laid out properly.

It appears to be a nuance of UICollectionViewLayout, rather than this library. Give it a shot and see if that fixes your issue.

**Comment by swapneshp-spaceo** [comment]
Is anyone found the solution, please let me know.
